hzToPeriod/periodToHz: use Ratio Natural

DigitalBrains1 opened the pull request:
This will more often give the intuitively correct solution, per the
principle of least surprise. Actual functionality should not be affected
by the exact rounding, and thus this should not significantly change the
behaviour of existing designs.

Example:

Before this commit:

```
>>> hzToPeriod 100_000
10000001
```
After this commit:
```
>>> hzToPeriod 100_000
10000000
```

\<edit\>
We would like `hzToPeriod` to satisfy the following:

- Results should be intuitive
- Results should match a type-level equivalent as proposed in issue #1253

The former can be done with `Double`, by rounding to the nearest integer. This means that if there is an exact integer solution, we give that. Before this commit we had:

```
>>> hzToPeriod 100_000
10000001
```

instead of the intuitive result that a 100 kHz clock corresponds to a 10 µs period. The off-by-one has no significant consequences for the circuit, clocks are inherently not precise. It is just not intuitive.

But if we also want the result to match a simply expressed type-level equivalent, this is not enough, because the slight errors introduced by `Double` might cause a fractional part of 0.5 to be rounded either up or down seemingly randomly.

So the most straightforward fix is to use `Ratio Natural` and `floor`. If the result is an integral number of picoseconds, it will give that, and for roundings, it will match the behaviour of the straightforward type-level equivalent of:

```
type HzToPeriod (freq :: Nat) = 1_000_000_000_000 `Div` freq
```
\</edit\>

* ~~_Does this need a Changelog entry?_ Observable behaviour changes minutely, e.g., clock periods in generated waveforms might be a picosecond shorter or longer, two clocks might accumulate a slightly different phase relation than before. None of this affects functionality, so I'm inclined to say, no, a Changelog entry is too verbose. It's definitely subjective, though.~~
* Should we backport?

christiaanb commented:
We probably want to use a custom rounding function that ties half to infinity instead of even, as it's easier to implement a type-level equivalent for that.

martijnbastiaan commented:
I kinda missed it in the (large!) discussions, but why don't we simply switch to `Rational` instead of `Double`? It would mitigate _any_ rounding error anyway, right?

christiaanb commented:
No, the error wasn't due to the number format, but due to use of `ceiling`

DigitalBrains1 commented:
I have an idea but no time to do so right away. I'll be back.

christiaanb commented on clash-prelude/src/Clash/Explicit/Signal.hs:
```diff
@@ -367,7 +367,7 @@ systemResetGen = resetGen
 -- __NB__: This function is /not/ synthesizable
 freqCalc :: Double -> Integer
-freqCalc = toInteger . hzToPeriod
+freqCalc = toInteger . hzToPeriod . realToFrac
```
You can just remove this function since we cannot backport this PR to the 1.2 branch.

DigitalBrains1 replied:
I'll do it tomorrow!

DigitalBrains1 replied:
Done

martijnbastiaan commented:
@DigitalBrains1 Should `periodToHz` output a `Ratio Natural`? Does this mean that this comment is obsolete:

```haskell
-- __NB__: This function is lossy. I.e., hzToPeriod . periodToHz /= id.
```

?

DigitalBrains1 commented:
Even if `periodToHz` output a `Ratio Natural`, it would still be lossy, because we round to whole picoseconds in the period.

I don't have a strong conviction, but to me, `Ratio` suggests exact numbers. And clock frequencies are inherently inexact. Even America's prime time and frequency standard, the atomic [NIST-F1](https://en.wikipedia.org/wiki/NIST-F1) clock only has an accuracy that is pretty much the same as the accuracy of a `Double` variable (cute coincidence!).

The only reason for `Ratio` here, for me, is to get equivalency with the type-level construction.

martijnbastiaan commented:
Alright. Wouldn't `periodToHz` have a similar type level equivalence though? (Also, I'm a sucker for symmetry :-))

DigitalBrains1 commented:
Yes, I just realised that that bit of documentation is now smelly anyway:
```
-- __NB__: This function is lossy. I.e., hzToPeriod . periodToHz /= id.
```
```
> :t hzToPeriod . periodToHz

<interactive>:1:14: error:
    • Couldn't match type ‘Double’
                     with ‘GHC.Real.Ratio GHC.Natural.Natural’
      Expected type: GHC.Natural.Natural
                     -> GHC.Real.Ratio GHC.Natural.Natural
        Actual type: GHC.Natural.Natural -> Double
    • In the second argument of ‘(.)’, namely ‘periodToHz’
      In the expression: hzToPeriod . periodToHz
```
So, yes, that is indeed not `id` by a long shot, but not because it's lossy :-)

Yeah, you could make a type-level `PeriodToHz` but I don't know if anybody is going to use it. I mean, you use `hzToPeriod` and `HzToPeriod` to tell Clash the frequency the circuit should run at. The reverse, have Clash tell you the frequency, isn't that useful, I think.

Anyway, I'm happy to change the other one to `Ratio` as well. I like symmetry too.

DigitalBrains1 commented:
I just realised the doc wasn't without issues.

Because now, unless I'm very mistaken, `hzToPeriod . periodToHz` is equal to the identity, but the juxtaposition is not. So I corrected the doc to say that `periodToHz . hzToPeriod` is the lossy one.

I feel like we're taking baby steps in this PR. But we'll get there :-).

DigitalBrains1 commented:
I fixed the doctest failure. Note how it was the representation, not the result that changed here:
```diff
@@ -1433,7 +1434,7 @@ hzToPeriod freq = floor ((1.0 / freq) / 1.0e-12)
 -- i.e. to calculate the clock frequency of a clock with a period of 5000 ps:
 --
 -- >>> periodToHz 5000
--- 2.0e8
+-- 200000000 % 1
 --
 -- __NB__: This function is /not/ synthesizable
 periodToHz :: Natural -> Ratio Natural
```

I also introduced a line break (actualy, a paragraph break) between the two NBs in the doc.
